README.md
# Reseptin arpoja
Sovellus auttaa käyttäjää päättämään mitä hän voisi kokata tai leipoa.
Käyttäjä kertoo halutessaan haluaako hän sovelluksen ehdottavan lounas reseptejä vai jälkiruokia.
Resepti valitaan satunnaisesta tiedoston käyttäjän antamien ehtojen rajoissa.
Käyttäjä voi halutessaan pyytää myös uutta reseptiä ja lisätä annetun reseptin loppuun oman huomautuksensa.


## Python versio
Sovellus on testattu Python-versiolla 3.8 eikä se välttämättä toimi vanhempien Python versioiden kanssa.


## Dokumentaatio

* [Release](https://github.com/evas3/ot-harjoitustyo/releases/tag/finalrelease)

* [Käyttöohje](https://github.com/evas3/ot-harjoitustyo/blob/main/dokumentaatio/kayttoohje.md)

* [Arkkitehtuuri](https://github.com/evas3/ot-harjoitustyo/blob/main/dokumentaatio/arkkitehtuuri.md)

* [Vaatimusmäärittey](https://github.com/evas3/ot-harjoitustyo/blob/main/dokumentaatio/vaatimusmaarittely.md)

* [Työaikakirjanpito](https://github.com/evas3/ot-harjoitustyo/blob/main/dokumentaatio/tyoaikakirjanpito.md)

* [Testaus](https://github.com/evas3/ot-harjoitustyo/blob/main/dokumentaatio/testaus.md)

* [Changelog](https://github.com/evas3/ot-harjoitustyo/blob/main/dokumentaatio/changelog.md)


## Asennus ja käyttö
Käyttäjän tulee ensin asentaa vaaditut riippuvuudet käyttäen seuraavaa komentoa:

	poetry install

Ohjelma käynnistyy seuraavalla komennolla:

	poetry run invoke start

Tämän jälkeen käyttäjä seuraa ohjelman käyttäjälle antamia ohjeita.
Ohjelmaa käytetään graafisen käyttöliittymän kautta. Käyttäjä siis painaa hiirellä nappia ja halutessaan kirjoittaa huomioille tarkoitettuun tekstikenttään.


## Muut komentorivikomennot
Testien suoritus seuraavalla komennolla:

	poetry run invoke test

Testikattavuusraportin generointi htmlcov-hakemistoon onnistuu komennolla:

	poetry run invoke coverage-report

Pylint määritelmien tarkastaminen toimii seuraavalla komennolla:

	poetry run invoke lint


dokumentaatio/changelog.md
# Muutokset

## Viikko 3
-Luotu luokka Recipes

-Luokkaan Recipes metodi welcome jolla ohjelma alkaa

-Luokkaan Recipes metodit sweet ja salty jotka palauttaa halutusta listasta satunnaisen jäsenen

-Luokkaan Recipes metodi redo jolla käyttäjä voi suorittaa ohjelman uudelleen

-Luokkaan Recipes metodi __str__ joka palauttaa listan jäsenen


## Viikko 4
-Arkkitehtuurikaavio

-Nimet englanniksi


## Viikko 5
-Korjattu pääohjelman toimintaa

-Ohjelma ajettavissa "poetry run invoke start" komennon avulla

-Reseptien printtaaminen tekstitiedostosta

-Pylint

-Release


## Viikko 6
-Docstring

-Pylint komento

-Tiedostojen polut muutettu toimiviksi

-Reseptit omassa kansiossaan

-Graaginen käyttöliittymä alustettu

-Korjattu pääohjelman toimintaa


## Viikko 7
-Testit voi ajaa "poetry run invoke test" komennolla

-Graafinen käyttöliittymä saatettu valmiiksi

-Palautettuun reseptiin voi lisätä oman huomion

-Lopullinen release

-Pylint virheitä korjattu


src/tests/resepti_test.py
import unittest
from recipes import Recipes
from tkinter import Tk
from graafinen_kayttoliittyma import UI


class TestRecipes(unittest.TestCase):
    def setUp(self):
        self.recipe = Recipes()

    def test_setup_sweet(self):
        self.assertEqual(len(self.recipe.dessert), 3)

    def test_setup_salty(self):
        self.assertEqual(len(self.recipe.meal), 3)

    def test_setup_test_indicator(self):
        self.assertEqual(self.recipe.test_indicator, 0)

    def test_welcome_input_1(self):
        self.recipe.welcome(1)
        self.assertEqual(self.recipe.test_indicator, 1)

    def test_welcome_input_2(self):
        self.recipe.welcome(2)
        self.assertEqual(self.recipe.test_indicator, 2)

    def test_welcome_input_(self):
        self.recipe.welcome(0)
        self.assertEqual(self.recipe.test_indicator, 0)

    def test_sweet_empty(self):
        self.recipe.dessert = []
        self.recipe.sweet()
        self.assertEqual(self.recipe.test_indicator, 3.2)

    def test_sweet(self):
        self.recipe.sweet()
        self.assertEqual(self.recipe.test_indicator, 3.1)

    def test_salty_empty(self):
        self.recipe.meal = []
        self.recipe.salty()
        self.assertEqual(self.recipe.test_indicator, 3.4)

    def test_salty(self):
        self.recipe.salty()
        self.assertEqual(self.recipe.test_indicator, 3.3)

    def test_redo_0(self):
        self.assertEqual(self.recipe.redo(0), "quit")

    def test_redo(self):
        self.assertEqual(self.recipe.redo("x", self.recipe.welcome(
            1)) or self.recipe.welcome(2) or self.recipe.welcome(0))


if __name__ == "__main__":
    unittest.main()


laskarit/viikko2/unicafe/src/tests/kassapaate_test.py
import unittest
from kassapaate import Kassapaate
from maksukortti import Maksukortti

class TestKassapaate(unittest.TestCase):
    def setUp(self):
        self.kassa = Kassapaate()
        self.kortti = Maksukortti(1000)

    def test_alustus_oikein(self):
        self.assertEqual(self.kassa.kassassa_rahaa, 100000)
        self.assertEqual(self.kassa.edulliset+self.kassa.maukkaat, 0)

    def test_edullinen_kateisella_riittavasti(self):
        palautus = self.kassa.syo_edullisesti_kateisella(500)
        self.assertEqual(self.kassa.kassassa_rahaa, 100240)
        self.assertEqual(palautus, 260)
        self.assertEqual(self.kassa.edulliset, 1)

    def test_maukas_kateisella_riittavasti(self):
        palautus = self.kassa.syo_maukkaasti_kateisella(500)
        self.assertEqual(self.kassa.kassassa_rahaa, 100400)
        self.assertEqual(palautus, 100)
        self.assertEqual(self.kassa.maukkaat, 1)

    def test_edullinen_kateisella_vahan(self):
        palautus = self.kassa.syo_edullisesti_kateisella(200)
        self.assertEqual(self.kassa.kassassa_rahaa, 100000)
        self.assertEqual(palautus, 200)
        self.assertEqual(self.kassa.edulliset, 0)

    def test_maukas_kateisella_vahan(self):
        palautus = self.kassa.syo_maukkaasti_kateisella(300)
        self.assertEqual(self.kassa.kassassa_rahaa, 100000)
        self.assertEqual(palautus, 300)
        self.assertEqual(self.kassa.maukkaat, 0)

    def test_edullinen_kortilla_riittavasti(self):
        self.assertEqual(self.kassa.syo_edullisesti_kortilla(self.kortti), True)
        self.assertEqual(self.kassa.edulliset, 1)
        self.assertEqual(str(self.kortti), "Kortilla on rahaa 7.60 euroa")

    def test_maukas_kortilla_riittavasti(self):
        self.assertEqual(self.kassa.syo_maukkaasti_kortilla(self.kortti), True)
        self.assertEqual(self.kassa.maukkaat, 1)
        self.assertEqual(str(self.kortti), "Kortilla on rahaa 6.00 euroa")

    def test_edullinen_kortilla_vahan(self):
        self.maksukortti = Maksukortti(200)
        self.assertEqual(self.kassa.syo_edullisesti_kortilla(self.maksukortti), False)
        self.assertEqual(self.kassa.edulliset, 0)
        self.assertEqual(str(self.maksukortti), "Kortilla on rahaa 2.00 euroa")

    def test_maukas_kortilla_vahan(self):
        self.maksukortti = Maksukortti(300)
        self.assertEqual(self.kassa.syo_maukkaasti_kortilla(self.maksukortti), False)
        self.assertEqual(self.kassa.maukkaat, 0)
        self.assertEqual(str(self.maksukortti), "Kortilla on rahaa 3.00 euroa")

    def test_kortille_rahaa(self):
        self.kassa.lataa_rahaa_kortille(self.kortti, 200)
        self.assertEqual(self.kassa.kassassa_rahaa, 100200)
        self.assertEqual(str(self.kortti), "Kortilla on rahaa 12.00 euroa")

    def test_kortille_neg_rahaa(self):
        self.kassa.lataa_rahaa_kortille(self.kortti, -200)
        self.assertEqual(self.kassa.kassassa_rahaa, 100000)
        self.assertEqual(str(self.kortti), "Kortilla on rahaa 10.00 euroa")

laskarit/viikko2/unicafe/src/tests/maksukortti_test.py
import unittest
from maksukortti import Maksukortti

class TestMaksukortti(unittest.TestCase):
    def setUp(self):
        self.maksukortti = Maksukortti(1000)

    def test_luotu_kortti_on_olemassa(self):
        self.assertNotEqual(self.maksukortti, None)

    def test_kortin_saldo_alussa(self):
        self.assertEqual(str(self.maksukortti), "Kortilla on rahaa 10.00 euroa")

    def test_rahan_lataaminen(self):
        self.maksukortti.lataa_rahaa(1000)
        self.assertEqual(str(self.maksukortti), "Kortilla on rahaa 20.00 euroa")

    def test_rahan_ottaminen_saldoa_tarpeeksi(self):
        self.assertEqual(self.maksukortti.ota_rahaa(1000), True)

    def test_rahan_ottaminen_saldoa_liianvahan(self):
        self.assertEqual(self.maksukortti.ota_rahaa(2000), False)


src/reseptit.py
import random

class Recipes():
    def __init__(self):
        self.meal = ["keittoresepti", "sopparesepti", "soossiresepti"]
        self.dessert = ["kakkuresepti", "piirasresepti", "tortturesepti"]
        self.tiedosto = "ainesosasia.txt"
        print("Itse et osaa tehdä hyviä päätöksiä -minä voin päättää reseptit puolestasi")
        print("Tekeekö mieli kokata vai leipoa? Paina entteriä jos et osaa päättää")

    def welcome(self, sweet_or_salty):
        if sweet_or_salty.lower() == "m":
            return reseptit.sweet()
        elif sweet_or_salty.lower() == "s":
           return reseptit.salty()
        elif sweet_or_salty == "":
            choises = [reseptit.sweet, reseptit.salty]
            return random.choice(choises)()
        else:
            print()
            print("Yritetäänpä uudelleen")
            return reseptit.welcome()

    def sweet(self):
        print(reseptit.__str__(self.dessert))
        with open (self.tiedosto) as tiedosto:
            for rivi in tiedosto:
                print(rivi)
        reseptit.redo()

    def salty(self):
        print(reseptit.__str__(self.meal))
        with open (self.tiedosto) as tiedosto:
            for rivi in tiedosto:
                print(rivi)
        reseptit.redo()

    def redo(self):
        print()
        print("Haetko kenties jotakin muuta?")
        again = input("Valitaanko uusi resepti? Paina y-kirjainta niin yritetään uudelleen.")
        if again.lower() == "y":
            reseptit.welcome()
        print("Ole hyvä ja näkemiin!")
        
    def __str__(self, list):
        return random.choice(list)

if __name__ ==  "__main__":
    reseptit = Recipes()
    sweet_or_salty = input("Makeaa (paina m-kirjainta) vai Suolaista(paina s-kirjainta): ")
    reseptit.welcome(sweet_or_salty)


src/redo_view.py
from tkinter import ttk, constants
import tkinter as tk


class RedoVieW:
    """Luokka 'valitaanko uudelleen' näkymää varten
    """

    def __init__(self, root, doable):
        """Alustaa näkymän.

        Args:
           root: ttk alustusta
           doable: kertoo jos joitakin reseptejä ei ole
        """

        self._root = root
        self._frame = None
        self.yes_or_no = 0
        self.doable = doable

        self._initialize()

    def pack(self):
        """Näyttää senhetkisen näkymän.
        """

        self._frame.pack(fill=constants.X)

    def destroy(self):
        """Tuhoaa senhetkisen näkymän.
        """

        self._frame.destroy()

    def _initialize(self):
        """Kuvaa redo näkymän tekstit, napit ja rakenteen.
        """

        self._frame = ttk.Frame(master=self._root)
        self.choise2 = tk.IntVar()

        if self.doable == "sweet":
            label1 = ttk.Label(
                master=self._frame, text="No more sweet options", font="Helvetica 16")
            label1.grid(row=0, column=0, columnspan=2, padx=10, pady=10)

        elif self.doable == "salty":
            label1 = ttk.Label(
                master=self._frame, text="No more salty options", font="Helvetica 16")
            label1.grid(row=0, column=0, columnspan=2, padx=10, pady=10)

        label = ttk.Label(
            master=self._frame, text="Shall we try another recipe?", font="Helvetica 16")
        label_2 = ttk.Label(
            master=self._frame, text="Select one and press 'OK' or just press 'OK' to quit",
            font="Helvetica 16")

        self.button_yes = ttk.Button(master=self._frame, text="Yes, continue",
                                     command=lambda: self._button2_clicked(1))
        self.button_no = ttk.Button(master=self._frame, text="No, quit",
                                    command=lambda: self._button2_clicked(0))
        self.button_ok2 = ttk.Button(
            master=self._frame, text="OK", command=lambda: self.choise2.set(self.yes_or_no))

        label.grid(row=1, columnspan=2, padx=10, pady=10)
        label_2.grid(row=2, columnspan=2, padx=10, pady=10)
        self.button_yes.grid(row=4, column=1, columnspan=1, sticky=(
            constants.E, constants.W), padx=10, pady=10)
        self.button_no.grid(row=4, column=0, columnspan=1, sticky=(
            constants.E, constants.W), padx=10, pady=10)
        self.button_ok2.grid(columnspan=2, row=5, sticky=(
            constants.E, constants.W), padx=10, pady=10)

    def _button2_clicked(self, answer):
        """Käsittelee napinpainalluksen.

        Args:
           answer: käyttäjän nappivalinta
        """
        style = ttk.Style()
        style.configure("clicked_button.TButton", background="green")
        style.configure("not_clicked.TButton")
        if answer == 0:
            self.button_no.configure(style="clicked_button.TButton")
            self.button_yes.configure(style="not_clicked.TButton")
            self.yes_or_no = 0

        if answer == 1:
            self.button_yes.configure(style="clicked_button.TButton")
            self.button_no.configure(style="not_clicked.TButton")
            self.yes_or_no = 1


src/recipe_view.py
from tkinter import ttk, constants, Text
import tkinter as tk


class RecipeView:
    """Luokka reseptien tulostamisnäkymää varten
    """

    def __init__(self, root, file):
        """Alustaa näkymän.

        Args:
           root: ttk alustusta
           file: luettava tiedosto
        """

        self._root = root
        self._frame = None
        self.recipe = file
        self.add = tk.IntVar()

        self._initialize()

    def pack(self):
        """Näyttää senhetkisen näkymän.
        """

        self._frame.pack(fill=constants.X)

    def destroy(self):
        """Tuhoaa senhetkisen näkymän.
        """

        self._frame.destroy()

    def _initialize(self):
        """Lukee tiedoston sisällön
        """

        self._frame = ttk.Frame(master=self._root)

        label0 = ttk.Label(
            master=self._frame, text="(Scroll down to see the whole recipe)", font="Helvetica 10")
        label = ttk.Label(
            master=self._frame, text="When done add a note to this recipe", font="Helvetica 16")
        label_2 = ttk.Label(
            master=self._frame, text="Write the note below and press 'OK' or just press 'OK'",
            font="Helvetica 16")
        self.note = ttk.Entry(master=self._frame)
        self.button_add = ttk.Button(
            master=self._frame, text="OK", command=lambda: self._button_clicked())

        text = Text(self._frame)
        with open(self.recipe, 'r') as file:
            text.insert("end", chars=file.read())

        text.grid(padx=10, pady=10)
        label0.grid(padx=10, pady=10)
        label.grid(padx=10, pady=10)
        label_2.grid(padx=10, pady=10)
        self.note.grid(sticky=(constants.E, constants.W), padx=10, pady=10)
        self.button_add.grid(
            sticky=(constants.E, constants.W), padx=10, pady=10)

    def _button_clicked(self):
        """Käsittelee napin painamisen ja kirjoittaa huomion.
        """

        self.add.set(1)
        note_value = self.note.get()
        if note_value != "":
            with open(self.recipe, 'a+') as file:
                file.write('\n' + "NOTE: " + note_value)
                file.close()


src/welcome_view.py
from tkinter import ttk, constants
import tkinter as tk


class WelcomeView:
    """Luokka graafisen käyttöliittymän aloitusnäkymälle.
    """

    def __init__(self, root):
        """Alustaa näkymän.

        Args:
           root: ttk alustusta
        """

        self._root = root
        self._frame = None
        self._final_choise = 0
        self._initialize()

    def pack(self):
        """Näyttää senhetkisen näkymän.
        """

        self._frame.pack(fill=constants.X)

    def destroy(self):
        """Tuhoaa senhetkisen näkymän.
        """

        self._frame.destroy()

    def _initialize(self):
        """Kuvaa aloitusnäkymän tekstit, napit ja rakenteen
        """

        self._frame = ttk.Frame(master=self._root)
        self.choise = tk.IntVar()
        style = ttk.Style()

        label = ttk.Label(
            master=self._frame, text="Don't know what to do? Would you like to bake or cook?",
            font="Helvetica 16")
        label_2 = ttk.Label(
            master=self._frame, text="Select one and press 'OK' or if not sure just press 'OK'",
            font="Helvetica 16")

        style.configure("clicked_button.TButton", background="green")
        style.configure("not_clicked.TButton")
        self.button_sweet = ttk.Button(
            master=self._frame, text="Bake", style="not_clicked.TButton",
            command=lambda: self._button_clicked("sweet"))
        self.button_salty = ttk.Button(
            master=self._frame, text="Cook", style="not_clicked.TButton",
            command=lambda: self._button_clicked("salty"))
        self.button_ok = ttk.Button(
            master=self._frame, text="OK", command=lambda: self.choise.set(self._final_choise))

        label.grid(columnspan=2, sticky=constants.W, padx=10, pady=10)
        label_2.grid(columnspan=2)
        self.button_sweet.grid(row=2, column=1, columnspan=1, sticky=(
            constants.E, constants.W), padx=10, pady=10)
        self.button_salty.grid(row=2, column=0, columnspan=1, sticky=(
            constants.E, constants.W), padx=10, pady=10)
        self.button_ok.grid(columnspan=2, row=3, sticky=(
            constants.E, constants.W), padx=10, pady=10)

    def _button_clicked(self, sweet_or_salty):
        """Käsittelee napinpainalluksen.

        Args:
           sweet_or_salty: käyttäjän nappivalinta
        """

        style = ttk.Style()
        style.configure("clicked_button.TButton", background="green")
        style.configure("not_clicked.TButton")

        if sweet_or_salty == "sweet":
            self.button_sweet.configure(style="clicked_button.TButton")
            self.button_salty.configure(style="not_clicked.TButton")
            self._final_choise = 1

        if sweet_or_salty == "salty":
            self.button_salty.configure(style="clicked_button.TButton")
            self.button_sweet.configure(style="not_clicked.TButton")
            self._final_choise = 2
            

src/graafinen_kayttoliittyma.py
from recipe_view import RecipeView
from welcome_view import WelcomeView
from redo_view import RedoVieW


class UI:
    """Luokka graafista käyttöliittymää varten.
    """

    def __init__(self, root):
        """Alustaa näkymän.

        Args:
            root: ttk alustusta
        """

        self._root = root
        self._current_view = None

    def tklinter_welcome(self):
        """Kutsuu ensimmäistä näkymää.
        """

        return self._show_welcome_view()

    def _show_welcome_view(self):
        """Ensimmäinen näkymä
           tervehtii käyttäjää.

        Returns:
           Nappulavalinnan arvon.
        """

        self._hide_current_view()

        self._current_view = WelcomeView(self._root)
        self._current_view.pack()

        self._current_view.button_ok.wait_variable(self._current_view.choise)
        return self._current_view.choise.get()

    def _hide_current_view(self):
        """Piilottaa nykyisen näkymän
        """

        if self._current_view:
            self._current_view.destroy()

        self._current_view = None

    def show_recipe_view(self, file):
        """Näyttää reseptinäkymän

        Args:
           file: luettava tiedosto
        """

        self._hide_current_view()

        self._current_view = RecipeView(self._root, file)
        self._current_view.pack()

        self._current_view.button_add.wait_variable(self._current_view.add)

    def show_redo_view(self, doable):
        """Näyttää 'yritetäänkö uudelleen' näkymän.

        Args:
           doable: kertoo jos reseptit ovat loppuneet

        Returns:
           Nappulavalinnan arvo.
        """

        self._hide_current_view()

        self._current_view = RedoVieW(self._root, doable)
        self._current_view.pack()

        self._current_view.button_ok2.wait_variable(self._current_view.choise2)
        return self._current_view.choise2.get()


src/recipes.py
import random
from tkinter import Tk
from graafinen_kayttoliittyma import UI


class Recipes():
    """Luokka jonka avulla päätetään reseptit.
    """

    def __init__(self):
        """Konstruktori luo listat resepteistä.
        """

        self.meal = ["./src/reseptit/nuudelit.txt",
                     "./src/reseptit/keitto.txt",
                     "./src/reseptit/hampurilaiset.txt"]
        self.dessert = ["./src/reseptit/mansikkakakku.txt",
                        "./src/reseptit/mustikkapiirakka.txt",
                        "./src/reseptit/kaaretorttu.txt"]
        self._test_indicator = 0


    def main(self):
        """Luo tkinter ikkunan
        """
        self.window = Tk()
        self.ui_window = UI(self.window)

        self.window.title("What to cook?")
        sweet_or_salty = self.ui_window.tklinter_welcome()
        self.welcome(sweet_or_salty)
        self.window.mainloop()

    def welcome(self, sweet_or_salty):
        """Toivottaa käyttäjän tervetulleeksi 
           määrittää valitaanko suolainen vai makea resepti.

        Args:
            sweet_or_salty: päätös kysymykseen yllä.
        """

        if sweet_or_salty == 1:
            self._test_indicator = 1
            self.sweet()
        elif sweet_or_salty == 2:
            self._test_indicator = 2
            self.salty()
        else:
            choises = [self.sweet, self.salty]
            random.choice(choises)()

    def sweet(self):
        """Tulostaa leivonnaisen reseptin jos se on.
           Vanhaa reseptiä ei enää ehdoteta.
        """

        if self.dessert:
            self._test_indicator = 3.1
            chosen_dessert = random.choice(self.dessert)
            self.ui_window.show_recipe_view(chosen_dessert)
            self.dessert.remove(chosen_dessert)
            self.redo("")
        else:
            self._test_indicator = 3.2
            self.redo("sweet")

    def salty(self):
        """Tulostaa ruoan reseptin jos on.
           Vanhaa reseptiä ei enää ehdoteta.
        """

        print()
        if self.meal:
            self._test_indicator = 3.3
            chosen_meal = random.choice(self.meal)
            self.ui_window.show_recipe_view(chosen_meal)
            self.meal.remove(chosen_meal)
            self.redo("")
        else:
            self._test_indicator = 3.4
            self.redo("salty")

    def redo(self, doable):
        """Kysyy käyttäjältä valitaanko uusi resepti.

        Args:
            doable: kertoo jos reseptit ovat loppuneet
        """

        quit_if_0 = self.ui_window.show_redo_view(doable)
        if quit_if_0 == 0:
            self.window.destroy()
            print()
            print("You're welcome!")
            return "quit"

        sweet_or_salty = self.ui_window.tklinter_welcome()
        return self.welcome(sweet_or_salty)


if __name__ == "__main__":
    recipes = Recipes()
    recipes.main()
